### ISSUE TITLE
Extend Gnuk ECC support (curve25519, secp256k1)

### ISSUE THREAD

**Issue by alex-nitrokey** [OpenSC/OpenSC#1715]
### Problem Description

The Gnuk/Nitrokey Start is a software implementation of the OpenPGP Card standard. It differs in some terms though. As these devices do support two curves not defined in the standard, it might be interesting to add these in OpenSC as well.

### Proposed Resolution
Add the curves for Gnuk devices only. This shall not be done [here](https://github.com/OpenSC/OpenSC/blob/19711d0a1301f0519a1348390bb996fd2925f766/src/libopensc/card-openpgp.c#L158), but in a switch-case when [actually added](https://github.com/OpenSC/OpenSC/blob/19711d0a1301f0519a1348390bb996fd2925f766/src/libopensc/card-openpgp.c#L636). Please note that the comment in said code part is not accurate anymore. Gnuk devices are recognized as such.

### Further work to do
I might implement the code myself in the future. Meanwhile this issue is a reminder for anyone eager to implement this and for others to may give head ups. For example: I do not know right now if `curve25519` and `secp256k1` are handled by OpenSC anyway etc.

**Comment by szszszsz** [comment]
Connected: https://github.com/OpenSC/OpenSC/pull/319

**Comment by popovec** [comment]
Hi, 

I have tested OpenSC with secp256k1 key - only ESDSA-SHA1 is working for me. MyEID card driver, OsEID card/simulation. OsEID simulation is already used to test OpenSC on travis. Example sign test with secp256k1 key  is already in OsEID-tool with small patch:

````
diff --git a/OsEID-tool b/OsEID-tool
index 9825ba7..abd2642 100755
--- a/OsEID-tool
+++ b/OsEID-tool
@@ -489,6 +489,7 @@ echo storing EC key
 cat <<EOP
 prime192v1 1.2.840.10045.3.1.1
 prime256v1 1.2.840.10045.3.1.7
+secp256k1 1.3.132.0.10
 EOP
 if [ $ECC_SIZE_MAX -ge 384 ]; then echo "secp384r1  1.3.132.0.34"; fi
 if [ $ECC_SIZE_MAX -ge 521 ]; then echo "secp521r1  1.3.132.0.35"; fi
````


OpenSC need to be patched too:
````
diff --git a/src/libopensc/card-myeid.c b/src/libopensc/card-myeid.c
index d4ed17c4..733b2331 100644
--- a/src/libopensc/card-myeid.c
+++ b/src/libopensc/card-myeid.c
@@ -108,6 +108,7 @@ static struct myeid_supported_ec_curves {
        {"secp384r1", {{1, 3, 132, 0, 34, -1}},         384},
        {"secp521r1", {{1, 3, 132, 0, 35, -1}},         521},
        {NULL, {{-1}}, 0},
+       {"secp256k1", {{1, 3, 132, 0, 10, -1}},         256},   /* OsEID */
 };
 
 static int myeid_get_info(struct sc_card *card, u8 *rbuf, size_t buflen);
@@ -251,6 +252,8 @@ static int myeid_init(struct sc_card *card)
                        if (card_caps.max_ecc_key_length >= ec_curves[i].size)
                                _sc_card_add_ec_alg(card, ec_curves[i].size, flags, ext_flags, &ec_curves[i].curve_oid);
                }
+               if (card->type == SC_CARD_TYPE_MYEID_OSEID)
+                       _sc_card_add_ec_alg(card,  ec_curves[5].size, flags, ext_flags, &ec_curves[5].curve_oid);
        }
 
        /* show supported symmetric algorithms */
````
Please note, there is problem to recognize NIST and other keys - for example pkcs15-tool -D generates similar output for secp256r1 and secp256k1:
````
Private EC Key [prime256v1]
        Object Flags   : [0x03], private, modifiable
        Usage          : [0x0C], sign, signRecover
        Access Flags   : [0x01], sensitive
        FieldLength    : 256
        Key ref        : 2 (0x02)
        Native         : yes
        Path           : 3f0050154b02
        Auth ID        : 01
        ID             : 169a6e46eb54302a71f826f3e6b6378faf1f2b69
        MD:guid        : 68c01f69-1ff6-05d0-b6de-33f5bc20a008

Private EC Key [secp256k1]
        Object Flags   : [0x03], private, modifiable
        Usage          : [0x0C], sign, signRecover
        Access Flags   : [0x01], sensitive
        FieldLength    : 256
        Key ref        : 3 (0x03)
        Native         : yes
        Path           : 3f0050154b03
        Auth ID        : 01
        ID             : 80729cd08ca02c3818d0ef4cc702ae666079aa70
        MD:guid        : 5671b19b-e108-62b6-6b01-43c0a6817257
````
I use `label` to recognize key type from user view, internally OsEID card uses special file type for secp256k1 key. 

pkcs11-tool -M does not allow us to check if secp256k1 is supported or  not:

````
$ pkcs11-tool -M
Using slot 0 with a present token (0x0)
Supported mechanisms:
  SHA-1, digest
  SHA224, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
  MD5, digest
  RIPEMD160, digest
  GOSTR3411, digest
  ECDSA, keySize={192,521}, hw, sign, other flags=0x1800000
  ECDSA-SHA1, keySize={192,521}, hw, sign, other flags=0x1800000
  ECDH1-COFACTOR-DERIVE, keySize={192,521}, hw, derive, other flags=0x1800000
  ECDH1-DERIVE, keySize={192,521}, hw, derive, other flags=0x1800000
  ECDSA-KEY-PAIR-GEN, keySize={192,521}, hw, generate_key_pair, other flags=0x1800000
  RSA-X-509, keySize={512,2048}, hw, decrypt, sign, verify, wrap, unwrap
  RSA-PKCS, keySize={512,2048}, hw, decrypt, sign, verify, wrap, unwrap
  SHA1-RSA-PKCS, keySize={512,2048}, sign, verify
  RSA-PKCS-PSS, keySize={512,2048}, hw, sign, verify, wrap, unwrap
  SHA1-RSA-PKCS-PSS, keySize={512,2048}, sign, verify
  RSA-PKCS-KEY-PAIR-GEN, keySize={512,2048}, generate_key_pair
  AES-ECB, keySize={128,256}, encrypt, decrypt, wrap, unwrap
  AES-CBC, keySize={128,256}, encrypt, decrypt, wrap, unwrap
  AES-CBC-PAD, keySize={128,256}, encrypt, decrypt, wrap, unwrap

````

And `pkcs11-tool -m ECDSA` or `-m ECDSA-SHA1` - result is same. 

If someone need  generated key, there is way to do this, but i have not found elegant way to pass curve name/OID  to card driver in actual opensc version. 

I welcome any suggestions to improve the current state of implementation of secp256k1 in OpenSC/OsEID.

**Comment by alex-nitrokey** [comment]
This issue will be kind of closed as soon as #1853 is merged. Currently there is no support for curve25519 in OpenSC at all (afaik), so there is no need to look for support on Gnuk for now.

As soon as support for curve25519 is implemented in OpenSC, [these lines](https://github.com/OpenSC/OpenSC/blob/7039a4bf6c5944c238ffffb9c28b318cf136d6cb/src/libopensc/card-openpgp.c#L183) needs to be uncommented only.
